keps/sig-node/0028-20180906-quotas-for-ephemeral-storage.md

```diff
@@ -0,0 +1,647 @@
+storage and inode consumption of volumes.  This mechanism suffers from
+a number of drawbacks:
+
+* It is slow.  If a volume contains a large number of files, walking
+  the directory can take a significant amount of time.  There has been
+  at least one known report of nodes becoming not ready due to volume
+  metrics: <https://github.com/kubernetes/kubernetes/issues/62917>
+* It is possible to conceal a file from the walker by creating it and
+  removing it while holding an open file descriptor on it.  POSIX
+  behavior is to not remove the file until the last open file
+  descriptor pointing to it is removed.  This has legitimate uses; it
+  ensures that a temporary file is deleted when the processes using it
+  exit, and it minimizes the attack surface by not having a file that
+  can be found by an attacker.  The following pod does this; it will
+  never be caught by the present mechanism:
+
+```yaml
+apiVersion: v1
+kind: Pod
+max:
+metadata:
+  name: "diskhog"
+spec:
+  containers:
+  - name: "perl"
+    resources:
+      limits:
+        ephemeral-storage: "2048Ki"
+    image: "perl"
+    command:
+    - perl
+    - -e
+    - >
+      my $file = "/data/a/a"; open OUT, ">$file" or die "Cannot open $file: $!\n"; unlink "$file" or die "cannot unlink $file: $!\n"; my $a="0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789"; foreach my $i (0..200000000) { print OUT $a; }; sleep 999999
+    volumeMounts:
+    - name: a
+      mountPath: /data/a
+  volumes:
+  - name: a
+    emptyDir: {}
+```
+* It is reactive rather than proactive.  It does not prevent a pod
+  from overshooting its limit; at best it catches it after the fact.
+  On a fast storage medium, such as NVMe, a pod may write 50 GB or
+  more of data before the housekeeping performed once per minute
+  catches up to it.  If the primary volume is the root partition, this
+  will completely fill the partition, possibly causing serious
+  problems elsewhere on the system.
+
+In many environments, these issues may not matter, but shared
+multi-tenant environments need these issues addressed.
+
+### Goals
+
+* Primary: improve performance of monitoring by using project quotas
+  in a non-enforcing way to collect information about storage
+  utilization.
+* Primary: detect storage used by pods that is concealed by deleted
+  files being held open.
+* Primary: this will not interfere with the more common user and group
+  quotas.
+* Stretch: enforce limits on per-volume storage consumption by using
+  enforced project quotas.  Each volume would be given an enforced
+  quota of the total ephemeral storage limit of the pod.
+
+### Non-Goals
+
+* Enforcing limits on total pod storage consumption by any means, such
+  that the pod would be hard restricted to the desired storage limit.
+
+## Proposal
+
+This proposal applies project quotas to emptydir volumes on qualifying
+filesystems (ext4fs and xfs with project quotas enabled).  Project
```

Comment:
What do we plan to do for other filesystems?  If we are changing the behavior for exceeding container limits (eviction vs "no space left on device"), ideally we want to keep the behavior the same across filesystems.  If that isn't possible, we should explicitly point it out.

Reply:
Exceeding container limits will still result in eviction.  And there's always the risk, even today, of running out of storage; we don't (and can't!) promise containers that they will have an unlimited amount of local storage to play with.

I'd be happy to support quotas on as many filesystems as we can, but if the underlying filesystem doesn't support project quotas, there's little we can do.

Reply:
> Exceeding container limits will still result in eviction

Is this because we are excluding the enforcement of container limits from the proposal, and leaving the current implementation as is?

Reply:
Is the behavior changing for empty-dirs (eviction-> "no space left on device")?

Reply:
Hitting the limit is still hitting the limit.  If you hit the limit and quickly got back under you'd avoid eviction, but again, no different from today.  If you hit the limit and stay there, the eviction code will pick up on it.

Consider setting the quota to one filesystem block more than the ephemeral limit (which I've considered doing).  If you write up to the quota limit and stop, you will definitely be above the ephemeral limit, so the eviction manager will toss you when it sees it.

Reply:
Also consider what happens if you have two emptydir volumes with a total limit of 1 MiB.  If you write 513 KiB to each you won't hit the quota on either, but you will still be evicted.

This proposal will certainly make "no space left on device" and similar failures more common, but they can still happen today.

Reply:
> If you hit the limit and stay there, the eviction code will pick up on it.

Ah, I see.  I was under the impression that pods that hit the disk quota limit would just stay there.  Might be worth clarifying that this is not replacing evictions for exceeding container limits (or empty-dir, for that matter).

Reply:
Adding an explicit non-goal that eviction will be eliminated.

Reply:
I think it's worth adding a section to clarify the what number to use when setting the quota, and how this interact with eviction. It's not very clear (at least to me) from reading the proposal.